x/fee_grant/internal/types/expiration.go

```diff
@@ -0,0 +1,134 @@
+package types
+
+import "time"
+
+// ExpiresAt is a point in time where something expires.
+// It may be *either* block time or block height
+type ExpiresAt struct {
+	Time   time.Time `json:"time" yaml:"time"`
+	Height int64     `json:"height" yaml:"height"`
+}
+
+// ExpiresAtTime creates an expiration at the given time
+func ExpiresAtTime(t time.Time) ExpiresAt {
+	return ExpiresAt{Time: t}
+}
+
+// ExpiresAtHeight creates an expiration at the given height
+func ExpiresAtHeight(h int64) ExpiresAt {
+	return ExpiresAt{Height: h}
+}
+
+// ValidateBasic performs basic sanity checks.
+// Note that empty expiration is allowed
+func (e ExpiresAt) ValidateBasic() error {
+	if !e.Time.IsZero() && e.Height != 0 {
+		return ErrInvalidDuration("both time and height are set")
+	}
+	if e.Height < 0 {
+		return ErrInvalidDuration("negative height")
+	}
+	return nil
+}
+
+// IsZero returns true for an uninitialized struct
+func (e ExpiresAt) IsZero() bool {
+	return e.Time.IsZero() && e.Height == 0
+}
+
+// FastForward produces a new Expiration with the time or height set to the
+// new value, depending on what was set on the original expiration
+func (e ExpiresAt) FastForward(t time.Time, h int64) ExpiresAt {
+	if !e.Time.IsZero() {
+		return ExpiresAtTime(t)
+	}
+	return ExpiresAtHeight(h)
+}
+
+// IsExpired returns if the time or height is *equal to* or greater
+// than the defined expiration point. Note that it is expired upon
+// an exact match.
+//
+// Note a "zero" ExpiresAt is never expired
+func (e ExpiresAt) IsExpired(t time.Time, h int64) bool {
+	if !e.Time.IsZero() && !t.Before(e.Time) {
+		return true
+	}
+	return e.Height != 0 && h >= e.Height
+}
+
+// IsCompatible returns true iff the two use the same units.
+// If false, they cannot be added.
+func (e ExpiresAt) IsCompatible(p Duration) bool {
+	if !e.Time.IsZero() {
+		return p.Clock > 0
+	}
+	return p.Block > 0
+}
+
+// Step will increase the expiration point by one Duration
+// It returns an error if the Duration is incompatible
+func (e ExpiresAt) Step(p Duration) (ExpiresAt, error) {
```

Comment:
why the `p`?
```suggestion
func (e ExpiresAt) Step(d Duration) (ExpiresAt, error) {
```

Reply:
p for Period, but I guess it should match type.